consensus/cometbft/service/prepare_proposal.go

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2024, Berachain Foundation. All rights reserved.
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package cometbft
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	ctypes "github.com/berachain/beacon-kit/consensus-types/types"
+	"github.com/berachain/beacon-kit/consensus/types"
+	"github.com/berachain/beacon-kit/primitives/math"
+	cmtabci "github.com/cometbft/cometbft/abci/types"
+)
+
+func (s *Service[LoggerT]) prepareProposal(
+	_ context.Context,
+	req *cmtabci.PrepareProposalRequest,
+) (*cmtabci.PrepareProposalResponse, error) {
+	startTime := time.Now()
+	defer s.telemetrySink.MeasureSince(
+		"beacon_kit.runtime.prepare_proposal_duration", startTime)
+
+	// CometBFT must never call PrepareProposal with a height of 0.
+	if req.Height < 1 {
+		return nil, fmt.Errorf(
+			"prepareProposal at height %v: %w",
+			req.Height,
+			errInvalidHeight,
+		)
+	}
+
+	// Always reset state given that PrepareProposal can timeout
+	// and be called again in a subsequent round.
+	s.prepareProposalState = s.resetState()
+	s.prepareProposalState.SetContext(
+		s.getContextForProposal(
+			s.prepareProposalState.Context(),
+			req.Height,
+		),
+	)
+
+	var slotData = types.NewSlotData[
+		ctypes.AttestationData,
+		ctypes.SlashingInfo,
+	](
+		math.Slot(req.GetHeight()),
+		nil,
+		nil,
+		req.GetProposerAddress(),
+		req.GetTime(),
+	)
+
+	// TODO: can we use the context from the request now that async is gone?
+	//nolint:contextcheck // see todo above
```

Comment:
is this comment still relevant? I don't think we can do that. But we may need to replace the base context with the one from the request, in order to properly handle potential cancellations from CometBFT

Reply:
haha, I wrote that before I knew that we were saving the state from the context, will remove comment